syntheseus/cli/search.py

```diff
@@ -0,0 +1,263 @@
+
+        alg = RetroStarSearch(**alg_kwargs)
+    elif config.search_algorithm == "mcts":
+        alg_kwargs.update(OmegaConf.to_container(config.mcts_config))
+        build_node_evaluator("value_function")
+        build_node_evaluator("reward_function")
+        build_node_evaluator("policy")
+
+        alg_kwargs["bound_function"] = lookup_by_name(mcts_base, alg_kwargs["bound_function_class"])
+        del alg_kwargs["bound_function_class"]
+
+        alg = MolSetMCTS(**alg_kwargs)
+    else:
+        raise NotImplementedError(f"Unsupported search algorithm {config.search_algorithm}")
+
+    # Prepare the output directory
+    results_dir_top_level = Path(config.results_dir)
+    timestamp = datetime.datetime.now().isoformat(timespec="seconds")
+    results_dir_current_run = results_dir_top_level / f"{config.model_class.name}_{str(timestamp)}"
+
+    logger.info("Setup completed")
+    for idx, smiles in enumerate(tqdm(search_targets)):
+        logger.info(f"Running search for target {smiles}")
+
+        if len(search_targets) == 1:
+            results_dir = results_dir_current_run
+        else:
+            results_dir = results_dir_current_run / str(idx)
+
+        results_dir.mkdir(parents=True, exist_ok=True)
+        logger.info(f"Outputs will be saved under {results_dir}")
+
+        alg.reset()
+        output_graph, _ = alg.run_from_mol(Molecule(smiles))
+        logger.info(f"Finished search for target {smiles}")
+
+        # Time of first solution (rxn model calls)
+        for node in output_graph.nodes():
+            node.data["analysis_time"] = node.data["num_calls_rxn_model"]
+        soln_time_rxn_model_calls = get_first_solution_time(output_graph)
+
+        # Time of first solution (wallclock)
+        for node in output_graph.nodes():
+            node.data["analysis_time"] = (
+                node.creation_time - output_graph.root_node.creation_time
+            ).total_seconds()
+        soln_time_wallclock = get_first_solution_time(output_graph)
+
+        stats = {
+            "index": idx,
+            "smiles": smiles,
+            "rxn_model_calls_used": alg.reaction_model.num_calls(),
+            "num_nodes_in_final_tree": len(output_graph),
+            "soln_time_rxn_model_calls": soln_time_rxn_model_calls,
+            "soln_time_wallclock": soln_time_wallclock,
+        }
+
+        logger.info(pformat(stats))
+
+        with open(results_dir / "stats.json", "wt") as f:
+            f.write(json.dumps(stats, indent=2))
+
+        if config.save_graph:
+            with open(results_dir / "graph.pkl", "wb") as f:
+                pickle.dump(output_graph, f)
+
+        if config.num_routes_to_plot > 0:
+            # Extract some synthesis routes in the order they were found
+            logger.info(f"Extracting up to {config.num_routes_to_plot} routes for analysis")
+
+            # TODO(kmaziarz): Add options to extract a diverse (or otherwise interesting) subset.
```

Comment:
In a future PR we can add `cli/analysis.py` and then have `cli/search.py` expose the options from `analysis.py` (but set the default analysis to something very lightweight, e.g. extracting a small diverse subset).